src/libraries/BytesParsing.sol

```diff
@@ -0,0 +1,1274 @@
+    bytes memory encoded,
+    uint offset,
+    uint length
+  ) internal pure returns (bytes memory ret, uint nextOffset) {
+    //bail early for degenerate case
+    if (length == 0)
+      return (new bytes(0), offset);
+
+    assembly ("memory-safe") {
+      nextOffset := add(offset, length)
+      ret := mload(freeMemoryPtr)
+
+      //Explanation on how we copy data here:
+      //  The bytes type has the following layout in memory:
+      //    [length: 32 bytes, data: length bytes]
+      //  So if we allocate `bytes memory foo = new bytes(1);` then `foo` will be a pointer to 33
+      //    bytes where the first 32 bytes contain the length and the last byte is the actual data.
+      //  Since mload always loads 32 bytes of memory at once, we use our shift variable to align
+      //    our reads so that our last read lines up exactly with the last 32 bytes of `encoded`.
+      //  However this also means that if the length of `encoded` is not a multiple of 32 bytes, our
+      //    first read will necessarily partly contain bytes from `encoded`'s 32 length bytes that
+      //    will be written into the length part of our `ret` slice.
+      //  We remedy this issue by writing the length of our `ret` slice at the end, thus
+      //    overwritting those garbage bytes.
+      let shift := and(length, 31) //equivalent to `mod(length, 32)` but 2 gas cheaper
+      if iszero(shift) {
+        shift := wordSize
+      }
+
+      let dest := add(ret, shift)
+      let end := add(dest, length)
+      for {
+        let src := add(add(encoded, shift), offset)
+      } lt(dest, end) {
+        src := add(src, wordSize)
+        dest := add(dest, wordSize)
+      } {
+        mstore(dest, mload(src))
+      }
+
+      mstore(ret, length)
+      //When compiling with --via-ir then normally allocated memory (i.e. via new) will have 32 byte
+      //  memory alignment and so we enforce the same memory alignment here.
+      mstore(freeMemoryPtr, and(add(dest, 31), not(31)))
+    }
+  }
+
+  function slice(
+    bytes memory encoded,
+    uint offset,
+    uint length
+  ) internal pure returns (bytes memory ret, uint nextOffset) {
+    (ret, nextOffset) = sliceUnchecked(encoded, offset, length);
+    checkBound(nextOffset, encoded.length);
+  }
+
+  function asAddressUnchecked(
+    bytes memory encoded,
+    uint offset
+  ) internal pure returns (address, uint) {
+    (uint160 ret, uint nextOffset) = asUint160(encoded, offset);
+    return (address(ret), nextOffset);
+  }
+
+  function asAddress(
+    bytes memory encoded,
+    uint offset
+  ) internal pure returns (address ret, uint nextOffset) {
+    (ret, nextOffset) = asAddressUnchecked(encoded, offset);
+    checkBound(nextOffset, encoded.length);
+  }
+
+  function asBoolUnchecked(
+    bytes memory encoded,
+    uint offset
+  ) internal pure returns (bool, uint) {
+    (uint8 ret, uint nextOffset) = asUint8(encoded, offset);
+    return (ret != 0, nextOffset);
```

Comment:
can we check that `ret == 0 || ret == 1` here? or equivalently `ret == ret & 1` which might be marginally cheaper

Reply:
Wouldn't `ret == ret & 1` return true for all odd numbers though, so it only half does what you're looking for. Agree though that it would be nice to enforce 0/1, good call

Reply:
hmm, I think it does what we want 🤔 are you thinking of `ret | 1`?

Reply:
Unless I'm being stupid I think that's right since it's a bitwise AND. If `a` is always 1, this little Remix test return true for `b` = 1,3,5,etc

```
pragma solidity 0.8.0;

contract Test {
    function testAnd(uint8 a, uint8 b) public pure returns (bool) {
        uint8 x = a & b;
        return (x != 0);
    }
}
```

Reply:
> You're suggesting we return ret == ret & 1 and forget about the !=0 part?

no no! I see the confusion now, what I'm suggesting is that we assert that `ret == ret & 1` then return what we're currently returning.

Reply:
Right I've gotcha now. So add in an `assert(ret == ret & 1)` before the return on line 99, sounds good to me!

Reply:
exactly, though thinking about it more, `assert(ret == 0 || ret == 1)` is a _lot_ more readable and costs marginally more gas 😅

Reply:
`assert` is the wrong choice since it should indicate situations that should never happen (implementation errors) and consumes all remaining gas. So it should be either a custom error or `require`.

Reply:
I'd propose the following implementation:

```
function asBoolUnchecked(
  bytes memory encoded,
  uint offset
) internal pure returns (bool, uint) {
  (uint8 val, uint nextOffset) = asUint8(encoded, offset);
  if (val & 0xfe != 0)
    revert InvalidBoolVal(val);
   
  bool ret;
  assembly ("memory-safe") {
    ret := val
  }
  return (ret, nextOffset);
}
```

this should translates to:
* 1 PUSH for the 0xfe constant
* 1 AND for &
* 2 ISZERO for != comparison
* 1 JUMPI

and the assembly section skips the 2 ISZERO opcods that the solidity compiler would otherwise throw in there to cleanly convert a uint8 into a bool

Reply:
> exactly, though thinking about it more, `assert(ret == 0 || ret == 1)` is a _lot_ more readable and costs marginally more gas 😅

I think in this part of the code we should strongly prioritize gas efficiency over readability. solc will turn the `||` here into another conditional jump because of short-circuiting which makes the entire thing a lot less gas and code efficient.

Reply:
fixing my sloppy oversight of using `asUint8` instead of `asUint8Unchecked` [here](https://github.com/wormhole-foundation/wormhole-solidity-sdk/pull/25)

Reply:
And as it turns out the inline assembly is not safe because solc is actually no cleaning up the upper 31 bytes.

Also fixed in the above PR...

My bad :crying_cat_face: